### PR TITLE
 Zonotope overapproximation of polytope given generator directions 

### PR DESCRIPTION
**Pull request by schillic**
See also #1175.

### CODE REVIEW

File: src/Approximations/overapproximate.jl [hunk 1]
```diff
@@ -1510,3 +1510,121 @@ function overapproximate(am::AbstractAffineMap{N, <:AbstractHyperrectangle{N}},
...
+nonnegativity constraints (last type) are not stated explicitly in [1].
+
+[1] Zonotopes as bounding volumes, L. J. Guibas et al, Proc. of Symposium on
+    Discrete Algorithms, pp. 803-812.
+"""
+# note: this set can generally be reduced
+function overapproximate(X::LazySet{N}, ::Type{<:Zonotope},
+                         dir::AbstractDirections{N};
+                         solver=default_lp_solver(N)) where {N<:Real}
+    function normalize_directions(dir::AD) where {AD}
```
Comment (mforets) [inline src/Approximations/overapproximate.jl]: seems like abstract directions are not concretely typed. although that is somehow unrelated to this PR, it would be nice to have them `AbstractDirections{N, VN}`  so we can dispatch on `VN`.

File: src/Approximations/overapproximate.jl [hunk 2]
```diff
@@ -1510,3 +1510,121 @@ function overapproximate(am::AbstractAffineMap{N, <:AbstractHyperrectangle{N}},
...
+            d_normalized = normalize(d)
+            if (d_normalized ∉ dirs) && (-d_normalized ∉ dirs)
+                push!(dirs, d_normalized)
+            end
+        end
+        return dirs
+    end
+
+    n = dim(X)
+    dirs = normalize_directions(dir)
```
Comment (mforets) [inline src/Approximations/overapproximate.jl]: this normalization should be optional, thinking that as use case for this function is to fix a template, normalize it (outside the function) and call `overapproximate` many times.

File: src/Approximations/overapproximate.jl [hunk 3]
```diff
@@ -1510,3 +1510,121 @@ function overapproximate(am::AbstractAffineMap{N, <:AbstractHyperrectangle{N}},
...
+### Algorithm
+
+We solve a linear program parametric in the vertices ``v_j`` of `X` and the
+directions ``d_k`` in `dir` presented in [1, Section 4.2] (adapting the notation
+to the one used in this library).
+
+```math
+    \\min \\sum_{k=1}^l c_k \\
+    s.t. \\
+    \\text{center} + \\sum_{k=1}^l b_{kj} * d_k = v_j \\quad \\forall j \\
```
Comment (mforets) [inline src/Approximations/overapproximate.jl]: can you use `c` for `\\text{center}` , in either case  define it below ?

File: src/Approximations/overapproximate.jl [hunk 4]
```diff
@@ -1510,3 +1510,121 @@ function overapproximate(am::AbstractAffineMap{N, <:AbstractHyperrectangle{N}},
...
+        end
+        return dirs
+    end
+
+    n = dim(X)
+    dirs = normalize_directions(dir)
+    l = length(dirs)
+    V = vertices_list(X)
+    m = length(V)
+    nvariables = l + n + l * m  # l 'c_k' + n 'p[i]' + lm 'b_lj'
```
Comment (mforets) [inline src/Approximations/overapproximate.jl]: what is `p[i]`? the coordinates of the new center?

File: src/LazySets.jl [hunk 5]
```diff
@@ -25,6 +25,40 @@ import .Assertions: activate_assertions, deactivate_assertions
...
+
+If the package is loaded, this function has no effect.
+Otherwise it prints an error message.
+
+### Algorithm
+
+This function uses `@assert` and hence loses its ability to print an error
+message if assertions are deactivated.
+"""
+function require(package::Symbol; fun_name::String="", explanation::String="")
```
Comment (mforets) [inline src/LazySets.jl]: is this change related to the PR? also, this check can also be done with a macro, an implementation is https://github.com/mforets/ReachabilityAnalysis.jl/blob/master/src/Initialization/init.jl#L62

File: src/Approximations/overapproximate.jl [hunk 6]
```diff
@@ -1510,3 +1510,121 @@ function overapproximate(am::AbstractAffineMap{N, <:AbstractHyperrectangle{N}},
...
+We solve a linear program parametric in the vertices ``v_j`` of `X` and the
+directions ``d_k`` in `dir` presented in [1, Section 4.2] (adapting the notation
+to the one used in this library).
+
+```math
+    \\min \\sum_{k=1}^l c_k \\
+    s.t. \\
+    \\text{center} + \\sum_{k=1}^l b_{kj} * d_k = v_j \\quad \\forall j \\
+    -c_k ≤ b_{kj} ≤ c_k \\quad \\forall k, j \\
+    c_k ≥ 0 \\quad \\forall k, j \\
```
Comment (mforets) [inline src/Approximations/overapproximate.jl]: not sure that i understand the notation in lines 1553 and 1554, because line 1554 says "for all j" but `c_k` has only 1 index (k)

File: src/Approximations/overapproximate.jl (hunk 1 above)
Comment (mforets) [inline src/Approximations/overapproximate.jl]: since some of `AbstractDirections` are normalized by construction, we can create an `IsNormalized` trait and use it.

Comment (schillic) [reply src/Approximations/overapproximate.jl]: Normalization here implies that you do not have the opposite direction. All our `AbstractDirections` contain opposite direction.

File: src/Approximations/overapproximate.jl (hunk 4 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl]: Yes. This is the notation from the paper.

File: src/LazySets.jl (hunk 5 above)
Comment (schillic) [reply src/LazySets.jl]: In the end it was not needed but if we want to use `require` in other modules then this is necessary, so I kept it.

File: src/Approximations/overapproximate.jl (hunk 6 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl]: Line 1554 is a copy-paste error.

File: src/Approximations/overapproximate.jl (hunk 3 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl]: `c` conflicts with `c_k`.

File: src/Approximations/overapproximate.jl (hunk 2 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl]: Currently there is no implementation satisfying the normalization. We can do that later.

File: src/Approximations/overapproximate.jl (hunk 3 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl]: I renamed `c_k` to `α_k`.

File: src/Approximations/overapproximate.jl (hunk 1 above)
Comment (mforets) [reply src/Approximations/overapproximate.jl]: yes; by construction we know which are opposite, we can use that info.

File: src/Approximations/overapproximate.jl (hunk 2 above)
Comment (mforets) [reply src/Approximations/overapproximate.jl]: you can at least save the step `d_normalized = normalize(d)`

File: src/LazySets.jl (hunk 5 above)
Comment (mforets) [reply src/LazySets.jl]: do you mean other submodules of LazySets? can't you just import the function in submodules?

File: src/Approximations/overapproximate.jl [hunk 7]
```diff
@@ -1510,3 +1510,123 @@ function overapproximate(am::AbstractAffineMap{N, <:AbstractHyperrectangle{N}},
...
+            for sign_b in N[1, -1]
+                A[r, k] = N(-1)  # - α_k
+                A[r, col_offset_b + j] = sign_b  # ± b_kj
+                r += 1
+            end
+        end
+        col_offset_b += m
+    end
+    lower_bounds = vcat(zeros(N, l), fill(N(-Inf), nvariables - l))
+    lp = linprog(obj, A, sense, b, lower_bounds, Inf, solver)
```
Comment (mforets) [inline src/Approximations/overapproximate.jl]: for the problem formulation, i would prefer that we use the JuMP interface which is much less verbose (dependency which is anyways loaded)

File: src/Approximations/overapproximate.jl (hunk 1 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl]: But a `IsNormalized` trait would still return `false` in all cases.

File: src/Approximations/overapproximate.jl (hunk 7 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl]: I tried that first and spent quite some time but could not get it to work. All other LPs in the library also do not use `JuMP` so I do not see why this is necessary now.

File: src/Approximations/overapproximate.jl (hunk 2 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl]: Only for box directions and this is a cheap operation so I would prefer to not complicate the code with conditions for one type.

File: src/LazySets.jl (hunk 5 above)
Comment (schillic) [reply src/LazySets.jl]: I did, but the function must be exported before it can be imported. But we load `Approximations.jl` before the function was defined in `init.jl`.

https://github.com/JuliaReach/LazySets.jl/blob/894f77bd2309a80e02d5be50acfb6dbc7c617947/src/LazySets.jl#L137-L157

File: src/Approximations/overapproximate.jl (hunk 2 above)
Comment (mforets) [reply src/Approximations/overapproximate.jl]: > Only for box directions

why? polar and spherical directions also have (2-norm) one. and custom directions may have it if one defines them.

> complicate the code with conditions for one type

hmm? we can dispatch on the directions type / the trait.

File: src/LazySets.jl (hunk 5 above)
Comment (mforets) [reply src/LazySets.jl]: > I did, but the function must be exported before it can be imported.

not sure that i understood. why do you want to `import`? but we can do `using LazySets: foo` inside the `Approximations` module as for other functions (eg. `_rtol`).

Comment (schillic) [reply src/LazySets.jl]: By *imported* I meant `using` of course.

> but we can do using LazySets: foo inside the Approximations module

Yes, but only after moving the function upward.

File: src/Approximations/overapproximate.jl [hunk 8]
```diff
@@ -1510,3 +1510,123 @@ function overapproximate(am::AbstractAffineMap{N, <:AbstractHyperrectangle{N}},
...
+
+The resulting zonotope has center `c` and generators `α_k · d_k`.
+
+Note that the first type of side constraints is vector-based and that the
+nonnegativity constraints (last type) are not stated explicitly in [1].
+
+[1] Zonotopes as bounding volumes, L. J. Guibas et al, Proc. of Symposium on
+    Discrete Algorithms, pp. 803-812.
+"""
+# note: this set can generally be reduced
```
Comment (mforets) [inline src/Approximations/overapproximate.jl:1565]: do you mean that it can be reduced without (or with a small) error?

Comment (schillic) [reply src/Approximations/overapproximate.jl:1565]: Huh, I do not understand this comment and it does not make any sense at the top of the function. I will remove it :sweat_smile:

File: src/Approximations/overapproximate.jl (hunk 2 above)
Comment (schillic) [reply src/Approximations/overapproximate.jl:1586]: I added a TODO in the code.

File: src/LazySets.jl (hunk 5 above)
Comment (schillic) [reply src/LazySets.jl]: I removed this change now as it was not needed.

**Comment by schillic** [comment]
I had to remove zero generators in the test. Should we do that by default in the function?
